### ISSUE TITLE
[PATCH] persisted.sob buglet

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7025]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)| @warner reported|
|-|-|
|Trac ID|trac#491|
|Type|defect|
|Created|2004-02-07 03:48:18Z|

Attachments:

* [mainmod.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/014/014a1aa3f7a2c6578a3fa48e3dd84f775925489f/453563399a175cb6be3722284c24df043c34c868.patch) (572 bytes) - added by warner on 2004-02-07 03:48:18Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__491 491
type__defect defect
reporter__warner warner
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1076125698000000 1076125698000000
changetime__1471145857069932 1471145857069932
version__ 
owner__jerub jerub
cc__warner cc__moshez cc__jerub
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
I don't have the failing case any more, but a few weeks back I noticed an
exception occur in _EverythingEphemeral (a NameError), when I made the usual
persist-from-__mail__ mistake.

The problem turned out to be that 'mainMod' is not actually available from
_EverythingEphemeral, as it is defined locally in load(), in a stack frame
somewhere above _EE but not accessible to it.

The fix I found was to declare mainMod as a global, but there might be a better
one (like adding it as an attribute to the _EE object). Patch attached for your
consideration.

cheers,
 -Brian
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @jerub|
|-|-|

By putting a raise ValueError in _EE.__getattr__, I've ascertained that the current testcases don't test __getattr__, it seems it isn't called, or is very rarely called.

It would be nice to get a failing testcase before applying this patch, but I think it should be applied.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub commented|
|-|-|

have renamed the local variable in the 'load' method, 'load' to '_load', because it's an identically named function with a different signature, we don't want it to look like recursion or polymorphism.

Added a unit test to verify the problem goes away.

branch is sob-491

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @jerub|
|-|-|

 * Docstring of `testMainMod` doesn't conform to PEP 257 (or PEP 8)
 * `testMainMod` isn't as clear as it should be.  Is there a way you can add an explicit, meaningful assertion?  
 * Good move on renaming `_load`
 * I reckon mainMod should be an attribute of `_EverythingEphemeral`

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

Also, it's not terribly clear why you need to set an attribute on __main__, nor why you need to set 'app' to None.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub commented|
|-|-|

setting 'app' to none is required otherwise this traceback will appear:
```

Traceback (most recent call last):
  File "/Users/stephen/src/Twisted/branches/sob-491/twisted/test/test_sob.py", line 141, in testEverythingEphermalSetattr
    sob.load('persisttest.mainmod', 'source')
  File "/Users/stephen/src/Twisted/branches/sob-491/twisted/persisted/sob.py", line 174, in load
    value = _load(fp)
  File "/Users/stephen/src/Twisted/branches/sob-491/twisted/persisted/aot.py", line 266, in unjellyFromSource
    raise ValueError("%s needs to define an 'app', it didn't!" % stringOrFile)
exceptions.ValueError: <open file 'persisttest.mainmod', mode 'r' at 0x13d8068> needs to define an 'app', it didn't!
```

And I've just exposed another bug, more patches coming.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jerub|
|-|-|

```print >>``` is forbidden by the coding standard.  Also, these files are all static.  Is there any reason to rewrite them on every test run?  Even if it is, the filenames need to change: there's no guarantee those names will be unique, the files should be created in a testmethod-specific directory, created with self.mktemp().

Most of the codebase uses """ for docstrings.  It'd be nice if it were the entire codebase.  They should also start and end with a newline so the trial misfeature that picks up text from docstrings doesn't get triggered.  Likewise for whitespace between arguments in calls - ```file('name','w')``` should be ```file('name', 'w')```.

Jamming state onto ```__main__``` should probably be avoided.  How about replacing the real ```__main__``` with a fake one for the duration of the test (just put it in sys.modules)?

I don't see why ```testEverythingEphermalGetattr``` puts ```assertEquals``` onto ```__main__```.  Nothing seems to use it.  Also the test method is misspelled (as are the other two, and the name in the docstrings).

mainMod should probably be an argument to ```_EverythingEphemeral.__init__``` instead of an attribute set on the instance after it is created.  Nice job getting rid of that global, though.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub set owner to @exarkun|
|-|-|

Excellent review, thankyou.

I don't have a good reason for putting the file generation into each of the test cases beyond wanting everything in one place... Would you like me to move this to static files?

I've made it use a temporary directory now for each of the files, got rid of ```print >>```, used """ for docstrings, put newlines in, replaced mutating __main__ with a fake main module, added a space after commas in function calls, got rid of the assertEquals in __main__, added an __init__ to _EE.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @jerub|
|-|-|

I think this is acceptable for merging now, but I still have a few notes:

 * ```EverythingEphemeral``` itself is still undocumented, so it's not clear what "works" means in the context of ```testEverythingEphemeralGetattr```.
 * Putting python code into static strings is generally a silly thing to do.  Putting them into individual files makes it easier to read with emacs, and harder to do things like accidentally leave out a newline when modifying them.  If you _are_ going to use static strings, at least put them together.  Compare:
```python
f = file(filename, 'w')
f.write('import __main__\n')
f.write('if __main__.testMainModGetattr != 1: raise AssertionError\n')
f.write('app = None\n')
f.close()
```
```python
file(filename, 'w').write("""
import __main__
if __main__.testMainModGetattr != 1:
    raise AssertionError
app = None
""")
```
   or better yet
```python
import __main__
if __main__.testMainModGetattr != 1:
    raise AssertionError
app = None
# ... somewhere else ...
sob.load(os.path.join(sibpath(__file__, "test_sob_stubs"),"main_assert.py"))
```
 * This isn't just for ```file.write```; any time you have code that looks like this
```python
foo = Foo()
foo.bar("a")
foo.bar("b")
# ...
foo.bar("y")
foo.bar("z")
foo.baz()
```
   something has probably gone wrong.
 * Note also that the code in the strings there violates the coding standard: no simple suites.  Instantiate your exceptions.

**Comment by twisted-trac** [comment]
|<img alt="stephen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stephen set status to `closed`|
|-|-|

(In [17325]) merge sob-491, fixes [#7025](https://github.com/twisted/twisted/issues/7025)

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|GitHub commented|
|-|-|

In changeset 92d5333beab83fadae175d73ec7fcbf6b00a4d84
```
#!CommitTicketReference repository="" revision="92d5333beab83fadae175d73ec7fcbf6b00a4d84"
Merge pull request #491 from twisted/8752-rodrigc-runner-py3

Author: hawkowl, rodrigc
Reviewer: markrwilliams
Fixes: #8752

Fix test_reporterDeprecations on Windows Python 3
```
